lib/Conversion/TritonGPUToLLVM/AssertOpToLLVM.cpp

```diff
@@ -35,11 +35,14 @@ struct AssertOpConversion : public ConvertOpToLLVMPattern<triton::AssertOp> {
       }
     }
     llAssert(op, condition, adaptor.getMessage(), rewriter);
-    // Add a barrier to avoid a race condition in case an assert is followed by
-    // an op that may trap if the assert condition is true. Since the tensor in
-    // those two operations may have different layout we need to make sure all
-    // the threads are done executing the assert before going to the next op.
-    barrier();
+    if (isa<RankedTensorType>(condition.getType())) {
+      // Add a barrier to avoid a race condition in case an assert is followed
+      // by an op that may trap if the assert condition is true. Since the
+      // tensor in those two operations may have different layout we need to
+      // make sure all the threads are done executing the assert before going to
+      // the next op.
```

Comment:
btw this highlights an interesting thing I noticed a while back. Sometimes we do the same computation in multiple different layouts when there are assert statements because the assertion path doesn't have a data flow path to the next layout anchor operation. 

It's not really the end of the world, but an interesting quirk of the remove layout conversion code.

Reply:
right, yes this is definitely a limitation as we haven't tried very hard to make assert efficient.